extension-workspace-reference.md

```diff
@@ -0,0 +1,82 @@
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  workspaceReferenceProvider?: boolean;
+}
+```
+
+#### Workspace Reference Request
+
+The workspace reference request is sent from the client to the server to export project-wide references to dependencies. That is, the response strictly returns references in the project to symbols defined in dependencies.
+
+_Request_
+* method: 'workspace/reference'
+* params: `WorkspaceReferenceParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace Reference Request.
+ */
+interface WorkspaceReferenceParams {
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace where the `symbol` has been referenced.
+	 */
+	reference: Location;
+
+	/**
+	 * Metadata describing the symbol that is being referenced. It is up to the
+	 * language server to define what exact data this object contains. In
+	 * general, the response includes the same information that `workspace/symbol`
+	 * would return, but that is only a guideline / not a hard constraint.
+	 *
+	 * The information does NOT always uniquely qualify the symbol. The caller
+	 * should effectively consider the returned information to be information
+	 * about the symbol which _generally_ (but NOT always) identifies a single
+	 * symbol.
+	 *
+	 * The following keys are reserved to have special meaning, but are entirely
+	 * optional:
+	 *
+	 * 	- `"name"` (same as `SymbolInformation.name`)
+	 * 	- `"kind"` (same as `SymbolInformation.kind`)
+	 * 	- `"file"` (same as `SymbolInformation.location.uri`)
+	 * 	- `"containerName"` (same as `SymbolInformation.containerName`)
+	 * 	- `"repo"`: A repository URI like `github.com/golang/go` or `bitbucket.org/jespern/django-piston`
+	 *
+	 * A language server is encouraged to include additional information about
+	 * the symbol. Specifically, any information that a user querying against
+	 * the returned information might be interested in. For example, with Go, a
+	 * user may wish to include or exclude vendored packages from their search
+	 * query, so the Go language server would include `"vendor": true` or `"vendor": false`
+	 * entry in the response which someone indexing this object could filter
+	 * based on. Another example would be including the canonical npm package
+	 * name such that, for example, a user could search for symbols defined in `"npm": "react"`.
+	 */
+	symbol: Object;
```

Comment:
Yes, I was thinking about this last night as well.. it will make extension/standardization of fields hard if they are mixed in with custom ones. In fact, I've already seen this with my PR to the Go LS which emits non-standard `dir` but obviously that could be standardized pretty easily.

What I'm now thinking is that we could have a separate type with 100% optional fields which represent the standardized values -- and then a `meta` property with other LS-specific properties..

I'll give this a shot.